lib/jormungandr/src/Cardano/Wallet/Network/Jormungandr/Api.hs

```diff
@@ -0,0 +1,116 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE TypeOperators #-}
+
+{-#  OPTIONS_GHC -fno-warn-orphans #-} -- for content types
+
+-- |
+-- Copyright: © 2018-2019 IOHK
+-- License: MIT
+--
+-- An specification for the Jörmungandr REST API.
+module Cardano.Wallet.Network.Jormungandr.Api
+    ( Api
+    , GetBlock
+    , GetTipId
+    , GetBlockDecendantIds
+    , PostSignedTx
+    , BlockId
+    , api
+    , SignedTx
+    ) where
+
+import Prelude
+
+import Cardano.Wallet.Binary.Jormungandr
+    ( FromBinary (..), runGet )
+import Cardano.Wallet.Primitive.Types
+    ( Block, Hash (..) )
+import Data.Binary.Get
+    ( getByteString )
+import Data.ByteArray.Encoding
+    ( Base (Base16), convertFromBase, convertToBase )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Text.Encoding
+    ( decodeUtf8 )
+import Servant.API
+    ( (:<|>)
+    , (:>)
+    , Capture
+    , Get
+    , MimeUnrender (..)
+    , NoContent
+    , Post
+    , QueryParam
+    , ReqBody
+    , ToHttpApiData (..)
+    )
+import Servant.API.ContentTypes
+    ( OctetStream, PlainText )
+
+import qualified Data.ByteString.Lazy as BL
+
+
+api :: Proxy Api
+api = Proxy
+
+type Api =
+    GetBlock :<|> GetTipId :<|> GetBlockDecendantIds :<|> PostSignedTx
+
+
+-- | Retrieve a block by its id.
+type GetBlock
+    = "api"
+    :> "v0"
+    :> "block"
+    :> Capture "blockHeaderHash" BlockId
+    :> Get '[OctetStream] Block
+
+-- | Retrieve a list of 'n' block decendant ids, sorted from closest to
+-- farthest.
+--
+-- There might also exist fewer than 'n' decendants.
+type GetBlockDecendantIds
+    = "api"
+    :> "v0"
+    :> "block"
+    :> Capture "blockId" BlockId
+    :> QueryParam "count" Int
+    :> Get '[OctetStream] [BlockId]
```

Comment:
I am not sure `OctetStream` is the best fit here. We actually know the format of the API we're talking to. So it's not just plain bytes. I'd suggest to define some custom format like `Jormungandr` or `BetterCBOR` or any name that fits to describe the format, and makes it clear that we're still expecting things in a particular format (whereas OctetStream conveys the idea of a stream of bytes without any particular structure). 

This will also prevent the orphan :+1:

Reply:
👍